python/cudf_polars/cudf_polars/dsl/expr.py

```diff
@@ -833,6 +833,24 @@ def do_evaluate(
 
 class TemporalFunction(Expr):
     __slots__ = ("name", "options", "children")
+    _temporal_extraction_function_map: ClassVar[dict[pl_expr.TemporalFunction, str]] = {
+        pl_expr.TemporalFunction.Year: "year",
+        pl_expr.TemporalFunction.Month: "month",
+        pl_expr.TemporalFunction.Day: "day",
+        pl_expr.TemporalFunction.WeekDay: "weekday",
+        pl_expr.TemporalFunction.Hour: "hour",
+        pl_expr.TemporalFunction.Minute: "minute",
+        pl_expr.TemporalFunction.Second: "second",
+        # millisecond/microsecond are implemented
+        # differently in polars, than seconds and above
+        # Instead of extracting the number of milliseconds/microseconds
+        # we must get the nanoseconds and then scale appropriately
+        # TODO: this should be possible
+        # pl_expr.TemporalFunction.Millisecond: "millisecond",
```

Comment:
I think polars implements millisecond/microsecond/nanosecond, by taking the sub-second part of the timestamp as nanoseconds and then dividing by 1000/1,000,000/etc.

I think this should be possible to match with pylibcudf.
Is there a way to get the sub-second part of a libcudf timestamp in nanoseconds?

Reply:
What about something like modding the underlying `int64` by `1e9`?

Reply:
libcudf has APIs for this. Please do not invent one. 😉 https://docs.rapids.ai/api/libcudf/stable/group__datetime__extract#ga78a831223016fd9b684d4b5db3500d56

Reply:
Doesn't get better than `extract_nanosecond_fraction` 🥇

Reply:
> libcudf has APIs for this. Please do not invent one. 😉 https://docs.rapids.ai/api/libcudf/stable/group__datetime__extract#ga78a831223016fd9b684d4b5db3500d56
> 
> `cudf::datetime::extract_nanosecond_fraction`

I don't think this does exactly what I want.

e.g. ``extract_nanosecond_fraction`` returns 890 for 1.234567890 seconds, but I want 234567890.

Reply:
can't we do the equivalent of 
```python
def get_subsecond_in_nanos(timestamp):
    casted = timestamp.astype(np.int64)  # Convert to integer (nanoseconds since epoch)
    remainder = casted % int(1e9)  # Get the remainder when divided by 1e9 (nanoseconds)
    return remainder
```
?

Reply:
If you want to implement a new API like `get_dt_field("total_nanos")` I would support that. I would avoid casting and modding, because that is putting "business logic" of datetime component extraction into something that should be a libcudf kernel.

Reply:
One reason you don't want to "cast and mod" is that not all datetimes have their "integral part" as nanoseconds. It looks like polars has `datetime[μs]` by default, which might convert to integer micros and not integer nanos. I'm not 100% sure about the exact polars behavior here but it is a real concern in pandas-land and in libcudf.

Reply:
https://github.com/rapidsai/cudf/pull/16500/commits/604720ba832f8e447b4b69662aee2d2eac5272f7 should implement the above